statistics/statistics_test.go

```diff
@@ -231,6 +232,87 @@ func (s *testStatisticsSuite) TestBuild(c *C) {
 	c.Check(int(count), Equals, 99999)
 }
 
+func (s *testStatisticsSuite) TestHistogramProtoConversion(c *C) {
+	ctx := mock.NewContext()
+	s.rc.Close()
+	tblCount, col, err := BuildIndex(ctx, 256, 1, ast.RecordSet(s.rc))
+	c.Check(err, IsNil)
+	c.Check(int(tblCount), Equals, 100000)
+
+	p := HistogramToProto(col)
+	h := HistogramFromProto(p)
+	c.Assert(col.NDV, Equals, h.NDV)
+	c.Assert(len(col.Buckets), Equals, len(h.Buckets))
+	for i, bkt := range col.Buckets {
+		c.Assert(bkt.Count, Equals, h.Buckets[i].Count)
+		c.Assert(bkt.Repeats, Equals, h.Buckets[i].Repeats)
+		c.Assert(bytes.Equal(bkt.LowerBound.GetBytes(), h.Buckets[i].LowerBound.GetBytes()), IsTrue)
+		c.Assert(bytes.Equal(bkt.UpperBound.GetBytes(), h.Buckets[i].UpperBound.GetBytes()), IsTrue)
+	}
+}
+
+func mockHistogram(lower, num int64) *Histogram {
+	h := &Histogram{
+		NDV: num,
+	}
+	for i := int64(0); i < num; i++ {
+		bkt := Bucket{
+			LowerBound: types.NewIntDatum(lower + i),
+			UpperBound: types.NewIntDatum(lower + i),
+			Count:      i + 1,
+			Repeats:    1,
+		}
+		h.Buckets = append(h.Buckets, bkt)
+	}
+	return h
+}
+
+func (s *testStatisticsSuite) TestMergeHistogram(c *C) {
+	sc := mock.NewContext().GetSessionVars().StmtCtx
+	bucketCount := 256
+	lh := mockHistogram(0, 0)
+	rh := mockHistogram(0, 1)
+	h, err := MergeHistograms(sc, lh, rh, bucketCount)
+	c.Assert(err, IsNil)
+	c.Assert(len(h.Buckets), Equals, 1)
+	c.Assert(h.NDV, Equals, int64(1))
+	c.Assert(h.Buckets[len(h.Buckets)-1].Count, Equals, int64(1))
+	cmp, err := h.Buckets[0].LowerBound.CompareDatum(sc, types.NewIntDatum(0))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	cmp, err = h.Buckets[len(h.Buckets)-1].UpperBound.CompareDatum(sc, types.NewIntDatum(0))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	lh = mockHistogram(0, 200)
+	rh = mockHistogram(200, 200)
+	h, err = MergeHistograms(sc, lh, rh, bucketCount)
+	c.Assert(err, IsNil)
+	c.Assert(h.NDV, Equals, int64(400))
+	c.Assert(len(h.Buckets), Equals, 200)
+	c.Assert(h.Buckets[len(h.Buckets)-1].Count, Equals, int64(400))
+	cmp, err = h.Buckets[0].LowerBound.CompareDatum(sc, types.NewIntDatum(0))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	cmp, err = h.Buckets[len(h.Buckets)-1].UpperBound.CompareDatum(sc, types.NewIntDatum(399))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	lh = mockHistogram(0, 200)
+	rh = mockHistogram(199, 200)
+	h, err = MergeHistograms(sc, lh, rh, bucketCount)
+	c.Assert(err, IsNil)
+	c.Assert(h.NDV, Equals, int64(399))
+	c.Assert(len(h.Buckets), Equals, 200)
+	c.Assert(h.Buckets[len(h.Buckets)-1].Count, Equals, int64(400))
+	cmp, err = h.Buckets[0].LowerBound.CompareDatum(sc, types.NewIntDatum(0))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	cmp, err = h.Buckets[len(h.Buckets)-1].UpperBound.CompareDatum(sc, types.NewIntDatum(398))
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
```

Comment:
There are three repeated tests, we can rewrite it to table driven test, it would be easier to read and add new test cases.